docs/gsl-intro.md

```diff
@@ -0,0 +1,327 @@
+
+
+## Iteration: "How do I iterate over a span?"
+
+A `span` is an encapsulated range, and so can be visited using a range-based `for` loop.
+
+Consider the implementation of a function like the `process_ints` that we saw in an earlier example. Visiting every object using a (pointer, length) pair requires an explicit index:
+
+~~~cpp
+void dangerous_process_ints(int* p, size_t n) {
+    for (auto i = 0; i < n; ++i) {
+        p[i] = next_character();
+    }
+}
+~~~
+
+A `span` supports range-`for` -- note this is zero-overhead and does not need to perform any range check, because the range-`for` loop is is known by construction not to exceed the range's bounds:
+
+~~~cpp
+void process_ints(span<int> s) {
+    for (auto& c : s) {
+        c = next_character();
+    }
+}
+~~~
+
+A `span` also supports normal iteration using `.begin()` and `.end()`.
+
+Note that you cannot compare iterators from different spans, even if they refer to the same array.
+
+An iterator is valid as long as the `span` that it is iterating over exists.
+
+
+## Element access: "How do I access a single element in a span?"
+
+Use `myspan[offset]` to subscript, or equivalently use `iter + offset` wheren `iter` is a `span<T>::iterator`. Both are range-checked.
+
+
+
+## Sub-spans: "What if I need a subrange of a span?"
+
+To refer to a sub-span, use `first`, `last`, or `subspan`.
+
+~~~cpp
+void process_ints(span<widget> s) {
+    if (s.length() > 10) {
+        read_header(s.first(10));   // first 10 entries
+        read_rest(s.subspan(10));   // remaining entries
+        // ...
+    }
+}
+~~~
+
+In rarer cases, when you know the number of elements at compile time and want to enable `constexpr` use of `span`, you can pass the length of the sub-span as a template argument:
+
+~~~cpp
+constexpr int process_ints(span<widget> s) {
+    if (s.length() > 10) {
+        read_header(s.first<10>());   // first 10 entries
+        read_rest(s.subspan<10>());   // remaining entries
+        // ...
+    }
+    return s.size();
+}
+~~~
+
+
+## span and STL: "How do I pass a span to an STL-style [begin,end) function?"
+
+Use `span::iterator`s. A `span` is iterable like any STL range.
+
+To call an STL `[begin,end)`-style interface, use `begin` and `end` by default, or other valid iterators if you don't want to pass the whole range:
+
+~~~cpp
+void f(span<widget> s) {
+    // ... 
+    auto found = find_if(s.begin(), s.end(), some_value);
+    // ... 
+}
+~~~
+
+If you are using a range-based algorithm such as from [Range-V3](https://github.com/ericniebler/range-v3), you can use a `span` as a range directly:
+
+~~~cpp
+void f(span<widget> s) {
+    // ... 
+    auto found = find_if(s, some_value); 
+    // ... 
+}
+~~~
+
+
+## Comparison: "When I compare `span<T>`s, do I compare the `T` values or the underlying pointers?"
+
+Comparing two `span<T>`s compares the `T` values. To compare two `span`s for identity, to see if they're pointing to the same thing, use `.data()`.
```

Comment:
I really dislike how github adds padding around code font, so that `span`s has so much space between the 'n' and the 's', but saying "`span` objects" everywhere probably isn't any better.